### DISCUSSION TITLE
Tidak ada residuals dari model? (Capstone ML)

### DISCUSSION THREAD

**Discussion by AGelfP** [teamalgoritma/community#638] [Q&A] [answered]
Dear Tim Algoritma,
Pertanyaan ini untuk Capstone kasus FnB.

Saya sudah sampai di langkah terakhir, yaitu uji asumsi autokorelasi dan normalitas. Namun saya tidak bisa melakukan kedua uji asumsi tersebut karena model saya tidak memiliki residual.

Sebagai latar belakang, terdapat 4 model time series yang telah saya buat. 3 dari model tersebut menggunakan fungsi exponential smoothing  `stlm()`, `tbats`, dan `bats` yang berasal dari package `forecast` . Model yang terakhir menggunakan fungsi `auto.msarima()` dari package `smooth`. Dokumentasi package tersebut dapat ditemukan [disini](https://search.r-project.org/CRAN/refmans/smooth/html/msarima.html). 

Berikut adalah bagaimana saya menuliskan kode-kode saya jika diperlukan

```
# model stlm
stlm_fnb <- fnb_train %>% 
  stlm() %>% 
  forecast(13*7)

# model tbats
tbat_model <- fnb_train %>% 
  tbats(use.box.cox = F,
        use.trend = T,
        use.damped.trend = T)

tbat_fnb <- forecast(tbat_model, h = 13*7)

# model bats
bats_model <- fnb_train %>% 
  bats(use.box.cox = F,
        use.trend = T,
        use.damped.trend = T)

bats_fnb <- forecast(bats_model, h = 13*7)

# model msarima
mod_msarima<- fnb_train %>% 
  auto.msarima()

sarima_fnb <- forecast(mod_msarima, h = 13*7)
```

Saya tidak menyangka bahwa model yang berasal dari package yang sama sekali tidak pernah disinggung tersebut justru menghasilkan nilai MAE yang paling rendah diantara 4 model pada saat proses cross-validation. Berikut adalah perbandingan dari keempat model.  (Catatan: Sepertinya nilai 0 adalah hasil dari saya menggantikan nilai NA menjadi 0 pada saat wrangling data)

```
# stlm
accuracy(stlm_fnb$mean, fnb_valid)

# tbat
accuracy(tbat_fnb$mean, fnb_valid)

# bats
accuracy(bats_fnb$mean, fnb_valid)

# SARIMA
accuracy(sarima_fnb$mean, fnb_valid)

                ME     RMSE      MAE  MPE MAPE      ACF1 Theil's U
Test set -5.440711 8.565184 7.057676 -Inf  Inf 0.4330558         0
                ME     RMSE      MAE  MPE MAPE      ACF1 Theil's U
Test set -4.812238 8.476199 6.818672 -Inf  Inf 0.5143262         0
                ME     RMSE      MAE MPE MAPE      ACF1 Theil's U
Test set -5.273133 8.378686 6.884808 NaN  Inf 0.4723932         0
                ME     RMSE      MAE  MPE MAPE      ACF1 Theil's U
Test set -1.495185 6.709143 5.147174 -Inf  Inf 0.4519131         0
```

Ketika saya uji ke leaderboard, hasilnya pun juga dibawah 0. Berikut ini adalah proses pengubahan nilai NA menjadi nilai prediksi model ke-4

```
submission_two <- submission %>% 
  mutate(visitor = floor(sarima_fnb$mean))

write.csv(submission_two, "Submission2_Alexander Gelfand Panjaya", row.names = F)
```

Dan ini adalah hasil leaderboard:

![Screenshot Bukti Leaderboard](https://user-images.githubusercontent.com/122977451/232210841-c1699a81-1f41-4d1c-9fc9-cbe0fef93ea5.png)

Jadi sampai pada titik tersebut, tidak ada masalah pada model saya. Namun ketika ingin melakukan uji asumsi yang sebelumnya disebutkan, tidak ada  opsi untuk memilih $residuals.

![image](https://user-images.githubusercontent.com/122977451/232211389-4b41a59a-a8ac-496a-9611-8180c59b782a.png)

Ketika saya cek secara manual menggunakan fungsi `checkresiduals()` dari package `forecast`, hasilnya malah menunjukkan pesan error

![image](https://user-images.githubusercontent.com/122977451/232212157-0f4adf0d-5ce8-488b-9f91-88e797e9bc9b.png)

Ketika menggunakan fungsi `residuals`, hasil justru null

![image](https://user-images.githubusercontent.com/122977451/232212085-6c076d4e-30a0-4339-9023-30fd3624426e.png)

Apakah saya telah melakukan kesalahan fatal dalam membentuk model saya karena membentuk model menggunakan package yang berbeda? Apakah ini akan menyebabkan pengurangan nilai juga? Mohon bantuannya.

**Comment by amaliasf** [comment, marked as answer]
Dear Pak @AGelfP 

Terimakasih atas pertanyaanya. Dan apresiasi kepada Pak Gelfand yang sudah eksplor pemodelan dengan package lain yg tidak dipelajari di kelas ^^

Berdasarkan referensi [Package Smooth](https://cran.r-project.org/web/packages/smooth/vignettes/smooth.html), untuk mengambil nilai residual dapat menggunakan code berikut: `residuals(ourModel)`

Namun pada screenshot yang Bapak lampirkan Bapak melakukan pengecekan residual terhadap hasil prediksi, bukan pada modelnya. Maka sepertinya code yang tepat adalah `residuals(mod_msarima)`

Boleh silahkan dicoba terlebih dahulu dan infokan kembali kalau masih terkendala. Semoga membantu.
